fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+        sigma: torch.Tensor,
+        sigma0: torch.Tensor,
+        epsilon: torch.Tensor,
+        is_smooth: bool = True,
+        is_var_computed: bool = True,
+        use_1sample_U: bool = True,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of deep-kernel MMD and std of deep-kernel MMD using merged data."""
+        X = Fea[0:len_s, :]  # fetch the sample 1 (features of deep networks)
+        Y = Fea[len_s:, :]  # fetch the sample 2 (features of deep networks)
+        X_org = Fea_org[0:len_s, :]  # fetch the original sample 1
+        Y_org = Fea_org[len_s:, :]  # fetch the original sample 2
+        L = 1  # generalized Gaussian (if L>1)
+        Dxx = self.Pdist2(X, X)
+        Dyy = self.Pdist2(Y, Y)
+        Dxy = self.Pdist2(X, Y)
+        Dxx_org = self.Pdist2(X_org, X_org)
+        Dyy_org = self.Pdist2(Y_org, Y_org)
+        Dxy_org = self.Pdist2(X_org, Y_org)
+        if is_smooth:
+            Kx = (1 - epsilon) * torch.exp(-((Dxx / sigma0) ** L) - Dxx_org / sigma) + epsilon * torch.exp(
+                -Dxx_org / sigma
+            )
+            Ky = (1 - epsilon) * torch.exp(-((Dyy / sigma0) ** L) - Dyy_org / sigma) + epsilon * torch.exp(
+                -Dyy_org / sigma
+            )
+            Kxy = (1 - epsilon) * torch.exp(-((Dxy / sigma0) ** L) - Dxy_org / sigma) + epsilon * torch.exp(
+                -Dxy_org / sigma
+            )
+        else:
+            Kx = torch.exp(-Dxx / sigma0)
+            Ky = torch.exp(-Dyy / sigma0)
+            Kxy = torch.exp(-Dxy / sigma0)
+
+        return self.h1_mean_var_gram(Kx, Ky, Kxy, is_var_computed, use_1sample_U)
+
+    def train_kernel(self, X: torch.Tensor, Y: torch.Tensor) -> None:
+        """Train the kernel."""
+
+        self.featurizer.train()
+        self.sigmaOPT.requires_grad = True
+        self.sigma0OPT.requires_grad = True
+        self.epsilonOPT.requires_grad = True
+
+        features = torch.cat([X, Y], 0)
+
+        # ------------------------------
+        #  Train deep network for MMD-D
+        # ------------------------------
+        # Initialize optimizer
+        self.optimizer_F.zero_grad()
+        # Compute output of deep network
+        model_output = self.featurizer(features)
+        # Compute epsilon, sigma and sigma_0
+        ep = torch.exp(self.epsilonOPT) / (1 + torch.exp(self.epsilonOPT))
+        sigma = self.sigmaOPT**2
+        sigma0_u = self.sigma0OPT**2
+        # Compute Compute J (STAT_u)
+        mmd_value_temp, mmd_var_temp, _ = self.MMDu(
+            Fea=model_output,
+            len_s=X.shape[0],
+            Fea_org=features.view(features.shape[0], -1),
+            sigma=sigma,
+            sigma0=sigma0_u,
+            epsilon=ep,
+        )
+        if mmd_var_temp is None:
+            raise AssertionError("Error: Variance of MMD is not computed. Please set is_var_computed=True.")
+        mmd_std_temp = torch.sqrt(mmd_var_temp + 10 ** (-8))
```

Comment:
I'm good with doing the sqrt here, since its how we form $\hat{J}_{\lambda}$. However, I'd move the lambda shift here into the formation of the estimate to collocate it better with the estimate calculation. So this would become something like 
```
mmd_std_estimate = torch.sqrt(mmd_var_estimate)
```